File: src/main/scala/com/hackerrank/algorithms/implementation/twosets/BetweenTwoSets.scala
package com.hackerrank.algorithms.implementation.twosets

object BetweenTwoSets {

  private def gcd(a: Long, b: Long): Long = {
    if (a == b) {
      a
    } else {
      if (a > b) {
        gcd(a - b, b)
      } else {
        gcd(a, b - a)
      }
    }
  }

  private def lcm(a: Long, b: Long): Long = {
    a * (b / gcd(a, b))
  }

  private def gcd(a: Array[Long]): Long = {
    if (a.isEmpty) {
      throw new IllegalArgumentException
    } else {
      if (a.length == 1) {
        a(0)
      } else {
        a.tail.foldLeft(a.head)(gcd)
      }
    }
  }

  private def lcm(a: Array[Long]): Long = {
    if (a.isEmpty) {
      throw new IllegalArgumentException
    } else {
      if (a.length == 1) {
        a(0)
      } else {
        a.tail.foldLeft(a.head)(lcm)
      }
    }
  }

  def getTotalX(a: Array[Long], b: Array[Long]): Long = {
    def count(cur: Long, low: Long, up: Long, res: Long): Long = {
      if (cur <= up || up % cur == 0) {
        if (up % cur == 0) {
          count(cur + low, low, up, res + 1)
        } else {
          count(cur + low, low, up, res)
        }
      } else {
        res
      }
    }

    val low = lcm(a)
    count(low, low, gcd(b), 0)
  }

  def main(args: Array[String]) {
    val sc = new java.util.Scanner(System.in);
    var n = sc.nextInt();
    var m = sc.nextInt();
    var a = new Array[Long](n);
    for (a_i <- 0 to n - 1) {
      a(a_i) = sc.nextInt();
    }
    var b = new Array[Long](m);
    for (b_i <- 0 to m - 1) {
      b(b_i) = sc.nextInt();
    }
    val total = getTotalX(a, b);
    println(total)
  }
}

File: src/main/java/com/hackerrank/algorithms/implementation/BetweenTwoSets.java
package com.hackerrank.algorithms.implementation;

import java.util.Arrays;
import java.util.Scanner;

public class BetweenTwoSets {

  static final int LIMIT = 101;

  private static int gcd(int a, int b) {
    while (b > 0) {
      int temp = b;
      b = a % b; // % is remainder
      a = temp;
    }
    return a;
  }

  public static int gcd(int[] a) {
    return Arrays.stream(a).reduce(a[0], BetweenTwoSets::gcd);
  }

  public static long lcm(long a, long b) {
    final int gcd = gcd((int) a, (int) b);
    long res = a / gcd * b;
    if (res > LIMIT) {
      return LIMIT;
    }
    return res;
  }

  public static long lcm(int[] a) {
    return Arrays.stream(a).mapToLong(Long::valueOf)
        .reduce(a[0], BetweenTwoSets::lcm);
  }

  public static boolean isPowerOfTwo(long b) {
    return (b & (b - 1)) == 0;
  }

  public static int solve(int[] a, int[] b) {
    // Complete this function
    final long lower = lcm(a);
    final long upper = gcd(b);
    int res = 0;
    for (long i = lower; i <= upper; i += lower) {
      if (upper % i == 0) {
        res++;
      }
    }
    return res;
  }

  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);
    int n = in.nextInt();
    int m = in.nextInt();
    int[] a = new int[n];
    for (int a_i = 0; a_i < n; a_i++) {
      a[a_i] = in.nextInt();
    }
    int[] b = new int[m];
    for (int b_i = 0; b_i < m; b_i++) {
      b[b_i] = in.nextInt();
    }
    int total = solve(a, b);
    System.out.println(total);
    in.close();
  }

}
